### PR TITLE
Devirtualize methods for -State combinator.

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Performance Improvement

Short description (up to few sentences):
State combinator mainly extends from the underlying function by one methods. Devirtualize it can lead to huge perf gain for certain use case.

```
Query:

select countMerge(d) from (select countState() d from numbers(1000000000));

Before:

Queries executed: 21.

localhost:9000, queries 21, QPS: 0.324, RPS: 324260324.149, MiB/s: 2473.910, result RPS: 0.324, result MiB/s: 0.000.

0.000%          3.020 sec.
10.000%         3.041 sec.
20.000%         3.047 sec.
30.000%         3.049 sec.
40.000%         3.052 sec.
50.000%         3.054 sec.
60.000%         3.056 sec.
70.000%         3.060 sec.
80.000%         3.083 sec.
90.000%         3.100 sec.
95.000%         3.130 sec.
99.000%         3.481 sec.
99.900%         3.560 sec.
99.990%         3.568 sec.


After:

Queries executed: 41.

localhost:9000, queries 41, QPS: 1.896, RPS: 1895701598.912, MiB/s: 14463.055, result RPS: 1.896, result MiB/s: 0.000.

0.000%          0.490 sec.
10.000%         0.512 sec.
20.000%         0.516 sec.
30.000%         0.521 sec.
40.000%         0.522 sec.
50.000%         0.526 sec.
60.000%         0.530 sec.
70.000%         0.538 sec.
80.000%         0.540 sec.
90.000%         0.544 sec.
95.000%         0.549 sec.
99.000%         0.571 sec.
99.900%         0.573 sec.
99.990%         0.574 sec.

```

### CODE REVIEW

**Comment by filimonov** [comment]
Some tests failed.

BTW - interesting how does it compare now with SimpleAggregateFunction ( https://github.com/ClickHouse/ClickHouse/pull/4629 )

**Comment by amosbird** [comment]
> BTW - interesting how does it compare now with SimpleAggregateFunction

tbh, I fail to see the correlation.

**Comment by filimonov** [comment]
SimpleAggregationFunction were introduced due to the big overhead of states processing of usual AggregationFunction. I'm wondering if were is a sense of using SimpleAggregationFunction after your change.

File: dbms/src/AggregateFunctions/AggregateFunctionGroupBitmap.cpp [hunk 1]
```diff
@@ -44,8 +44,9 @@ AggregateFunctionPtr createAggregateFunctionBitmapL2(const std::string & name, c
 
     const DataTypeAggregateFunction& datatype_aggfunc = dynamic_cast<const DataTypeAggregateFunction&>(*argument_type_ptr);
     AggregateFunctionPtr aggfunc = datatype_aggfunc.getFunction();
-    argument_type_ptr = aggfunc->getArgumentTypes()[0];
-    AggregateFunctionPtr res(createWithUnsignedIntegerType<AggregateFunctionTemplate, AggregateFunctionGroupBitmapData>(*argument_type_ptr, argument_type_ptr));
+    auto sub_argument_type_ptr = aggfunc -> getArgumentTypes()[0];
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupBitmap.cpp]: Style.

**Pull request closed without merging**